src/mongo_logger.ts

```diff
@@ -0,0 +1,184 @@
+  defaultSeverity: SeverityLevel;
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  maxDocumentLength: number;
+  /** Destination for log messages. Must be 'stderr' or 'stdout'. Defaults to 'stderr'. */
+  logDestination: string;
+}
+
+/** @internal */
+export class MongoLogger {
+  componentSeverities: Record<MongoLoggableComponent, SeverityLevel>;
+  maxDocumentLength: number;
+  logDestination: Writable;
+
+  constructor(options: MongoLoggerOptions) {
+    // TODO(NODE-4849): add filepath and Writable support
+    this.logDestination =
+      options.logDestination.toLowerCase() === 'stdout' ? process['stdout'] : process['stderr'];
+    this.componentSeverities = options;
+    this.maxDocumentLength = options.maxDocumentLength;
+  }
+
+  /* eslint-disable @typescript-eslint/no-unused-vars */
+  /* eslint-disable @typescript-eslint/no-empty-function */
+  emergency(component: any, message: any): void {}
+
+  alert(component: any, message: any): void {}
+
+  critical(component: any, message: any): void {}
+
+  error(component: any, message: any): void {}
+
+  warn(component: any, message: any): void {}
+
+  notice(component: any, message: any): void {}
+
+  info(component: any, message: any): void {}
+
+  debug(component: any, message: any): void {}
+
+  trace(component: any, message: any): void {}
+
+  /**
+   * Merges options set through environment variables and the MongoClient, preferring environment
+   * variables when both are set, and substituting defaults for values not set. Options set in
+   * constructor take precedence over both environment variables and MongoClient options.
+   *
+   * @remarks
+   * When parsing component severity levels, invalid values are treated as unset and replaced with
+   * the default severity.
+   *
+   * @param envOptions - options set for the logger from the environment
+   * @param clientOptions - options set for the logger in the MongoClient options
+   * @returns a MongoLoggerOptions object to be used when instantiating a new MongoLogger
+   */
+  static resolveOptions(
+    envOptions: MongoLoggerEnvOptions,
+    clientOptions: MongoLoggerMongoClientOptions
+  ): MongoLoggerOptions {
+    const defaultSeverity =
+      parseSeverityFromString(envOptions.MONGODB_LOG_ALL) ?? SeverityLevel.OFF;
+
+    return {
+      command: parseSeverityFromString(envOptions.MONGODB_LOG_COMMAND) ?? defaultSeverity,
+      topology: parseSeverityFromString(envOptions.MONGODB_LOG_TOPOLOGY) ?? defaultSeverity,
+      serverSelection:
+        parseSeverityFromString(envOptions.MONGODB_LOG_SERVER_SELECTION) ?? defaultSeverity,
+      connection: parseSeverityFromString(envOptions.MONGODB_LOG_CONNECTION) ?? defaultSeverity,
+      defaultSeverity,
+      maxDocumentLength: parseMaxDocumentLength(envOptions.MONGODB_LOG_MAX_DOCUMENT_LENGTH),
+      logDestination: parseLogDestination(envOptions.MONGODB_LOG_PATH, clientOptions.mongodbLogPath)
```

Comment:
- Is this going to be the only option configurable via `clientOptions`?
- My understanding from the logging slack channel discussion was that we’ve settled on `clientOptions` winning over `envOptions`, right?

Reply:
I think the design that this PR was based on did not get updated as the result of those conversations; I think we should support all options via the client, so we probably want to decide on appropriate option names for that, but I think we can do it in a separate ticket for ease of review - @andymina can you make sure to file the follow - up?

Reply:
I filed NODE-4878